### PR TITLE
Exercise 4

### PR DESCRIPTION
**Pull request by jobegrabber**
Review for lichtow: https://github.com/Systems-Development-and-Frameworks/lichtow/pull/9#pullrequestreview-548129733

### CODE REVIEW

File: backend/src/authorisation.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { allow, rule, shield } from 'graphql-shield'
```
Comment (powlaa) [inline backend/src/authorisation.js]: Didn't know there was already an allow from graphql-shield 👍 
Now we also used it in our permissions 😄

File: backend/src/resolvers.js [hunk 2]
```diff
@@ -1,13 +1,34 @@
...
+      const post = dataSources.db.getPost(id)
+      if (post.author.id !== userId) throw new ForbiddenError('May not delete a post of another user!')
+      return dataSources.db.deletePost(id)
+    },
+    upvotePost: (_, { id }, { dataSources, userId }) => dataSources.db.upvotePost(id, userId),
+    downvotePost: (_, { id }, { dataSources, userId }) => dataSources.db.downvotePost(id, userId),
+    signup: (_, { name, email, password }, { dataSources, getUserAuthenticationToken }) => {
+      if (password.length < 8) return new UserInputError('Password must be at least 8 characters long!')
+      return bcrypt.hash(password, bcryptSaltRounds)
+        .then((passwordHash) => getUserAuthenticationToken(dataSources.db.createUser(name, email, passwordHash)))
```
Comment (noeljns) [inline backend/src/resolvers.js]: Nice, you earned a star
> ⭐ For an implemented signup feature.

File: backend/src/resolvers.js [hunk 3]
```diff
@@ -1,13 +1,34 @@
...
+    },
+    login: (_, { email, password }, { dataSources, getUserAuthenticationToken }) => {
+      const user = dataSources.db.getUserByEmail(email)
+      return bcrypt.compare(password, user.passwordHash)
+        .then((isEqual) => isEqual
+          ? getUserAuthenticationToken(user.id)
+          : new UserInputError('Invalid password provided!'))
+    }
```
Comment (noeljns) [inline backend/src/resolvers.js]: I like the getUserAuthenticationToker method! A star
> ⭐ For an implemented login feature using JWT.

File: backend/src/authorisation.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+const permissions = shield({
+  Query: {
+    posts: allow,
+    users: isAuthenticated
+  },
+  Mutation: {
+    createPost: isAuthenticated,
+    deletePost: isAuthenticated,
+    upvotePost: isAuthenticated,
+    downvotePost: isAuthenticated,
```
Comment (powlaa) [inline backend/src/authorisation.js:27]: You could use a fallbackRule for this so any Mutation or Query you add will use this rule by default.

File: backend/src/db.js [hunk 5]
```diff
@@ -6,38 +6,65 @@ export class InMemoryNewsDS extends DataSource {
...
     this.users = new Map()
   }
 
-  createUser (name) {
-    if (this.users.has(name)) throw new Error(`User with name ${name} already exists!`)
-    this.users.set(name, new User(name))
+  createUser (name, email, passwordHash) {
+    if (this.hasUserWithEmail(email)) {
+      throw new Error(`User with email ${email} already exists!`)
+    }
```
Comment (noeljns) [inline backend/src/db.js]: Great, a star
> ⭐ For unique email address validation.

File: backend/src/resolvers.js [hunk 6]
```diff
@@ -1,13 +1,34 @@
...
+      if (password.length < 8) return new UserInputError('Password must be at least 8 characters long!')
+      return bcrypt.hash(password, bcryptSaltRounds)
+        .then((passwordHash) => getUserAuthenticationToken(dataSources.db.createUser(name, email, passwordHash)))
+    },
+    login: (_, { email, password }, { dataSources, getUserAuthenticationToken }) => {
+      const user = dataSources.db.getUserByEmail(email)
+      return bcrypt.compare(password, user.passwordHash)
+        .then((isEqual) => isEqual
+          ? getUserAuthenticationToken(user.id)
+          : new UserInputError('Invalid password provided!'))
```
Comment (noeljns) [inline backend/src/resolvers.js]: Nice implementation, another star
> ⭐ For password validation.

File: backend/src/server.js [hunk 7]
```diff
@@ -16,13 +20,23 @@ const db = new InMemoryNewsDS()
...
+              ]
+            }).map(source => source.document),
+            resolvers
+          }),
+          permissions
+        ),
+      dataSources: () => ({ db }),
+      context: ({ req }) => ({
+        getUserAuthenticationToken: (userId) => opts.authentication.getUserAuthenticationToken(userId),
+        userId: opts.authentication.checkUserAuthentication(req)
```
Comment (noeljns) [inline backend/src/server.js:38]: Another star
> ⭐ For assigning the authenticated user to a post in write and upvote mutations.
using getUserAuthenticationToken()

**Review by noeljns — CHANGES_REQUESTED** [review]
Overall a great implementation. As you already mentioned some tests are currently missing though.

File: backend/src/server.spec.js [hunk 8]
```diff
@@ -137,75 +158,87 @@ describe('queries', () => {
       `
     })
 
-    it('calls getUsers', async () => {
-      db.getUsers = jest.fn(() => [])
-      await usersQuery()
-      expect(db.getUsers).toHaveBeenCalledTimes(1)
-    })
+    describe('for unauthenticated users', () => {
```
Comment (powlaa) [inline backend/src/server.spec.js:165]: Good test cases for unauthenticated users 👍

**Review by powlaa — APPROVED** [review]
Looks really good overall 👍

File: backend/src/resolvers.js [hunk 9]
```diff
@@ -1,13 +1,89 @@
...
+        if (error instanceof PostIdNotFoundError) return error
+        throw error
+      }
+    },
+    signup: (_, { name, email, password }, { dataSources, getUserAuthenticationToken }) => {
+      if (password.length < 8) return new TooShortPasswordError()
+      if (dataSources.db.hasUserWithEmail(email)) return new EmailAlreadyExistsError()
+      return bcrypt.hash(password, bcryptSaltRounds)
+        .then((passwordHash) => getUserAuthenticationToken(dataSources.db.createUser(name, email, passwordHash)))
+    },
```
Comment (medizinmensch) [inline backend/src/resolvers.js:66]: :star: For an implemented signup   feature.
:star: For password validation.
:star:   For unique email address validation.

File: backend/src/resolvers.js [hunk 10]
```diff
@@ -1,13 +1,89 @@
...
+        user = dataSources.db.getUserByEmail(email)
+      } catch (error) {
+        if (error instanceof UserEmailNotFoundError) return error
+        throw error
+      }
+      return bcrypt.compare(password, user.getPasswordHash())
+        .then((isEqual) => isEqual
+          ? getUserAuthenticationToken(user.id)
+          : new InvalidPasswordError())
+    }
```
Comment (medizinmensch) [inline backend/src/resolvers.js:79]: :star: For an implemented login   feature using JWT.

File: backend/src/server.spec.js [hunk 11]
```diff
@@ -297,237 +351,456 @@ describe('mutations', () => {
...
+        errors: [error]
+      } = await signupMutation('TestUser', 't@t.de', 'someLongEnoughPassword')
+      expect(data).toBeNull()
+      expect(error.message).toBe(new EmailAlreadyExistsError().message)
     })
   })
 
-  describe('downvotePost', () => {
-    const downvotePostMutation = (id, voterName) => mutate({
+  describe('login', () => {
```
Comment (medizinmensch) [inline backend/src/server.spec.js:764]: :star:   For a software-tested login feature.

File: backend/src/server.spec.js [hunk 12]
```diff
@@ -297,237 +351,456 @@ describe('mutations', () => {
...
+                title: 'Hot new news',
+                votes: 99
+              }
+            }
+          })
+      })
+    })
+  })
+
+  describe('signup', () => {
```
Comment (medizinmensch) [inline backend/src/server.spec.js:698]: :star:   For a software-tested signup feature.

**Review by medizinmensch — APPROVED** [review]
Hi Jonas!

Sorry for the late review. It somehow slipped through. 

It looks very good. It's very easy to review your code because it's structure, minimal and easy to read. So thanks for that!

Additionally to what :star: you got below, you get the following:

:star:   For requesting a review and reviewing another team's PR according to the   instructions.
:star: For a successful rebase   on homework/main.
:star:   For not having security issues according to instructions.

That makes 10/10 :star: in total.

Proud

**Pull request merged**